internal/server/controllers/subscriptionController.go

```diff
@@ -0,0 +1,48 @@
+package controllers
+
+import (
+	"errors"
+	"github.com/gin-gonic/gin"
+	"gses4_project/internal/apperrors"
+	"gses4_project/internal/models"
+	"gses4_project/internal/services"
+	"net/http"
+)
+
+type ISubscriptionService interface {
+	Subscribe(email string) (*models.Email, error)
+}
+
+type SubscriptionController struct {
+	SubscriptionService ISubscriptionService
+}
+
+func NewSubscriptionController() *SubscriptionController {
+	return &SubscriptionController{
+		SubscriptionService: services.NewSubscriptionService(),
+	}
+}
+
+func (c *SubscriptionController) Subscribe(ctx *gin.Context) {
+	var subscriptionData models.Email
+	if err := ctx.ShouldBindJSON(&subscriptionData); err != nil {
+		ctx.JSON(http.StatusBadRequest, gin.H{
+			"error": "Couldn't bind request",
+		})
+		return
+	}
+
+	subscription, err := c.SubscriptionService.Subscribe(subscriptionData.Email)
+
+	if err != nil {
+		var httpErr *apperrors.HttpError
+		if errors.As(err, &httpErr) {
```

Comment:
This also applies here. The `SubscriptionService` could return the `SubscriptionAlreadyExists` error with the implementation of the `Error() string` function, in the controller you would check what kind of error it is and decide what to return to the client.